app/models/event.rb

```diff
@@ -19,4 +19,53 @@ class Event < ApplicationRecord
       )
     end
   end
+
+  def timeslots_with_ranking
+    timeslots_with_score = add_score_to_timeslot(timeslots)
+
+    sorted_timeslots = timeslots_with_score.sort do |a, b|
+      b.score <=> a.score
+    end
+
+    add_rank_to_timeslots sorted_timeslots
+  end
+
+  def add_score_to_timeslot(timeslots)
+    timeslots.map do |timeslot|
+      score = timeslot.preferences.reduce(0) do |sum, pref|
+        sum + preference_score(pref.preference_type)
+      end
+
+      timeslot.define_singleton_method(:score) { score }
+      timeslot.define_singleton_method(:rank) { nil }
+      timeslot
+    end
+  end
+
+  def add_rank_to_timeslots(timeslot_array)
+    timeslot_array.each_with_index.map do |timeslot, i|
+      if i < 3
+        ranking = 1 + i
+        timeslot.define_singleton_method(:rank) { ranking }
+      end
+      timeslot
+    end
+  end
+
+  def preference_score(preference)
+    return 1 if preference == 'tentative'
```

Comment:
I think a switch statement is also more readable here.  Typically you use implicit returns in ruby unless you are exiting a function early.
```ruby
    case preference
    when 'tentative' then 1
    when 'available' then 2
    when 'preferred' then 3
    when 'not_available' then -100
    else 0
    end
```

Reply:
👍